docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,343 @@
+metadata:
+  name: example-rule
+  namespace: default
+spec:
+  drain:
+    # behavior can be Drain or Skip
+    behavior: Drain 
+    # order can only be set if behavior == Drain
+    # Pods with higher order are drained later
+    order: 100
+  machines:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    clusterSelector:
+      matchLabels:
+      matchExpressions:
+  pods:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    namespaceSelector:
+      matchLabels:
+      matchExpressions:
+```
+
+Notes:
+
+* All selectors are of type `metav1.LabelSelector`
+* In order to allow expressing selectors with AND and OR predicates, both `machines` and `pods` have a list of 
+  selectors instead of a single selector
+
+##### Example: Exclude Pods from drain
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: skip-pods
+  namespace: default
+spec:
+  drain:
+    behavior: Skip
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchExpressions:
+      - key: app
+        operator: In
+        values:
+        - example-app1
+        - example-app2
+        - example-app3
+    namespaceSelector:
+      matchLabels:
+        kubernetes.io/metadata.name: example-namespace
+```
+
+##### Example: Drain order
+
+(related to [CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024))
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: portworx
+  namespace: default
+spec:
+  drain:
+    behavior: Drain
+    order: 100
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchLabels:
+        app: portworx
+```
+
+#### Drain labels
+
+We propose to introduce new labels to allow workload cluster admins/users to define drain behavior 
+for Pods. These labels would be either added directly to Pods or indirectly via Deployments, StatefulSets, etc.
+The labels will take precedence over `MachineDrainRules` specified in the management cluster.
+
+* `cluster.x-k8s.io/drain: skip`
+
+Initially we also considered adding a `cluster.x-k8s.io/drain-order` label. But we're not entirely sure about it 
+yet as someone who has access to the workload cluster (or maybe only specific Pods) would be able to influence the 
+drain order of the entire cluster, which might lead to problems. The skip label is safe in comparison because it 
+only influences the drain behavior of the Pod that has the label.
+
+#### Node drain behavior
+
+The following describes the new algorithm that decides which Pods should be drained and in which order.
+
+* List all Pods on the Node
+* For each Pod (first match applies):
+  * If the Pod is a DaemonSet Pod or a static Pod
+    * \=\> use `behavior: Skip`
+  * If the Pod has `cluster.x-k8s.io/drain: skip`
+    * \=\> use `behavior: Skip`
+  * If there is a matching `MachineDrainRule`
+    * \=\> use `behavior` and `order` from the first matching `MachineDrainRule` (based on alphabetical order)
```

Comment:
(Thinking out loud a bit here) Why alphabetically sorting the MDRs? This will likely lead to folks prefixing their rules with numbers, is that what we want?

```
01_MyDrainRuleIsKing
02_MyDrainRuleWasKingUntil01WasCreated
```

That said, I'm not sure I can think of a better way to achieve this.

If you tried to restrict the matching so that it always created disjoint sets, that may pass once, but may not pass later if new pods are added, so that's also fragile.

You could add an MDR priority in the spec and use a validation to ensure the priority is always unique

I guess the benefit of alphabetical is that kubectl will output them in priority order automatically?

I'd be interested to see other conversation that lead to the alphabetical decision, documenting that decision within the doc may be useful for the future (I don't see it in here yet?)

Reply:
Basically we were thinking:
* we need some criteria to decide which one is used
* restricting matching is impossible (you don't know all labels of all Pods upfront and they can also change)
* adding priority and enforcing that they are unique is possible, but probably not entirely race condition free and adding "another" prioritiy field on top of "order" is probably confusing

Reply:
Added to the proposal